### PR TITLE
Fix bug when an index appears in two expressions

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105055
* #102611
* #104987
* #104975
* __->__ #104886

We were not adding the bounds to `replacement_vars` in this case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

### CODE REVIEW

File: torch/utils/_sympy/value_ranges.py [hunk 1]
```diff
@@ -83,6 +83,19 @@ def __contains__(self, x):
         x = simple_sympify(x)
         return sympy_generic_le(self.lower, x) and sympy_generic_le(x, self.upper)
 
+    def tighten(self, other):
```
Comment (ezyang) [inline torch/utils/_sympy/value_ranges.py:86]: What exactly does tighten do?

Comment (lezcano) [reply torch/utils/_sympy/value_ranges.py:86]: i.e. given two bounds, it returns the AND of them.

Comment (eellison) [reply torch/utils/_sympy/value_ranges.py:86]: add docstring / type annotation ?

File: torch/_inductor/bounds.py [hunk 2]
```diff
@@ -87,9 +87,10 @@ def set_indirect(self, old, new):
     def get_index(self, name):
         expr = self.loop_body.indexing_exprs[name]
         bound = self.replacement_vals.get(expr)
-        if bound is not None:
-            return bound
-
-        bound = bound_sympy(expr, self.replacement_vals)
+        if bound is None:
```
Comment (eellison) [inline torch/_inductor/bounds.py:90]: when would we have None bound as opposed to `ValueRanges.unknown()`

Comment (lezcano) [reply torch/_inductor/bounds.py:90]: we're doing a `get()`, so when the key is not in the dictionary yet.

**Review by eellison — APPROVED** [review]

File: torch/_inductor/bounds.py [hunk 3]
```diff
@@ -87,9 +87,10 @@ def set_indirect(self, old, new):
...
+            bound = bound_sympy(expr, self.replacement_vals)
+        # The following assertion is true at the time of this writing
+        # We don't assert is as to not execute bound_sympy when bound is not None
+        # assert bound is None or bound == bound_sympy(expr, self.replacement_vals)
         self.replacement_vals[name] = bound
         return bound
```
Comment (Skylion007) [inline torch/_inductor/bounds.py:96]: Couldn't most of this method just be reduced to this one liner using set_default https://www.w3schools.com/python/ref_dictionary_setdefault.asp?
```suggestion
        return self.replacement_vals.setdefault(name, bound_sympy(expr, self.replacement_vals))
```

Comment (Skylion007) [reply torch/_inductor/bounds.py:96]: Actually, nvm that would always evaluate bound_sympy unfortunately which may be undesirable...

Comment (lezcano) [reply torch/_inductor/bounds.py:96]: That would eagerly execute `bound_sympy`, and that's exactly what we're trying to avoid.

**Review by Skylion007 — APPROVED** [review]

**Pull request closed without merging**